xmtp_id/src/associations/signature.rs

```diff
@@ -52,3 +59,182 @@ impl Clone for Box<dyn Signature> {
+        contract_address: String,
+        block_height: u64,
+    ) -> Self {
+        Erc1271Signature {
+            signature_text,
+            signature_bytes,
+            contract_address,
+            block_height,
+        }
+    }
+}
+
+impl Signature for Erc1271Signature {
+    fn recover_signer(&self) -> Result<MemberIdentifier, SignatureError> {
+        // TODO: Verify signature first
+        Ok(self.contract_address.clone().into())
+    }
+
+    fn signature_kind(&self) -> SignatureKind {
+        SignatureKind::Erc1271
+    }
+
+    fn bytes(&self) -> Vec<u8> {
+        self.signature_bytes.clone()
+    }
+
+    fn to_proto(&self) -> SignatureProto {
+        SignatureProto {
+            signature: Some(SignatureKindProto::Erc1271(Erc1271SignatureProto {
+                contract_address: self.contract_address.clone(),
+                block_height: self.block_height as i64,
+                signature: self.bytes(),
+            })),
+        }
+    }
+}
+
+#[allow(dead_code)]
+#[derive(Clone)]
+pub struct InstallationKeySignature {
+    signature_text: String,
+    signature_bytes: Vec<u8>,
+}
+
+impl InstallationKeySignature {
+    pub fn new(signature_text: String, signature_bytes: Vec<u8>) -> Self {
+        InstallationKeySignature {
+            signature_text,
+            signature_bytes,
+        }
+    }
+}
+
+impl Signature for InstallationKeySignature {
+    fn recover_signer(&self) -> Result<MemberIdentifier, SignatureError> {
+        todo!()
+    }
+
+    fn signature_kind(&self) -> SignatureKind {
+        SignatureKind::InstallationKey
+    }
+
+    fn bytes(&self) -> Vec<u8> {
+        self.signature_bytes.clone()
+    }
+
+    fn to_proto(&self) -> SignatureProto {
+        SignatureProto {
+            signature: Some(SignatureKindProto::InstallationKey(
+                RecoverableEd25519SignatureProto {
+                    bytes: self.bytes(),
+                },
+            )),
+        }
+    }
+}
+
+#[allow(dead_code)]
+#[derive(Clone)]
+pub struct LegacyDelegatedSignature {
+    // This would be the signature from the legacy key
+    legacy_key_signature: RecoverableEcdsaSignature,
+    signed_public_key: xmtp_proto::xmtp::message_contents::SignedPublicKey,
+}
+
+impl LegacyDelegatedSignature {
+    pub fn new(
+        legacy_key_signature: RecoverableEcdsaSignature,
+        signed_public_key: xmtp_proto::xmtp::message_contents::SignedPublicKey,
+    ) -> Self {
+        LegacyDelegatedSignature {
+            legacy_key_signature,
+            signed_public_key,
+        }
+    }
+}
+
+impl Signature for LegacyDelegatedSignature {
+    fn recover_signer(&self) -> Result<MemberIdentifier, SignatureError> {
+        // TODO: Two steps needed here:
+        // 1. Verify the RecoverableEcdsaSignature and make sure it recovers to the public key specified in the `signed_public_key`
+        // 2. Verify the wallet signature on the `signed_public_key`
+        // Return the wallet address
```

Comment:
In case it's helpful, for your next PR - there should be an implementation of this already that you can copy/paste/refactor, same with the todo's above

Reply:
ethers crate supports this and I'm working on [this issue](https://github.com/xmtp/libxmtp/issues/636).

btw, if anyone knows how to convert bytes(Vec<u8>, [u8] etc.) to/from different signature types(ethers, k256 etc.), that would be so helpful!

Reply:
I'd suggest writing a `TryFrom<...>` implementation for _our_ types, and if their libraries already have `TryFrom` implemented you can use that internally. That way you don't have to worry about constraints around implementing traits for foreign types.